### ISSUE TITLE
Database Unreadable in Many Other Applications

### ISSUE THREAD

**Issue by mbway** [keepassxreboot/keepassxc#2541]
I have a password database which was originally created in Keepass2 but which I now manage with KeePassXC (2.3.1) and recently something seems to have happened to the header which prevents the database from being read by KeepassX, Keepass2Android and KeePassDroid. The database can be read perfectly fine by Keepass2 and KeepassXC. Running 'Repair Database' from KeePassXC reports no errors.

Opening with Keepass2Android I get the error:
```
An error occurred: The file header is
corrupted. Less data than expected
could be read from the file.
```
Opening with KeePassDroid I get the error:
```
invalid database
```
opening in KeePassx gives an error:
```
Unable to open database.
Unsupported KeePass database version
```

I was going to post a bug report to Keepass2Android before realising the the other two also fail to read the database. Since a fresh database created by KeepassXC can also be read by these applications I assume there is a problem with KeePassXC.

I suspected that the cause was a problematic entry, I had recently added a few entries with various file attachments for example, but I made a copy of the database and removed every entry, changed the master password, the key derivation function and removed the keyfile and the problem remains.

I have tried opening the database in Keepass2 and saving, but the problem remains. KeepassX gave the same error, Keepass2Android gave a different error about the file being corrupted. It suggested saving without compression but this didn't fix the issue. KeePassDroid also gives a different error, now saying `invalid header type: 16`, after opening and saving again in Keepass2 without compression, KeePassDroid now says `invalid header type: -117`.

## Expected Behavior
The database should open correctly in these other applications.

## Current Behavior
The database is interpreted as invalid by these other applications.

## Possible Solution
I am unsure what exactly has changed about the database but I can provide additional information about the faulty database if shown how to extract it. If it is possible to cleanse all personal information from the database then I can provide the blank database with no entries but which has the same problems.

## Steps to Reproduce
I have been unable to reproduce by creating a new database and adding entries with file attachments etc. I can provide additional information if required.

## Context
I use KeepassXC on my laptop but I require my passwords on my phone as well, so I sync the database to my phone occasionally (always one way laptop -> phone, so none of the other applications have ever written to the database). None of the apps I use can read the database any-more.

## Debug Info
KeePassXC - Version 2.3.1
Revision: 2fcaeea

Libraries:
- Qt 5.9.5
- libgcrypt 1.8.1

Operating system: Ubuntu 18.04.1 LTS
CPU architecture: x86_64
Kernel: linux 4.15.0-42-generic

Enabled extensions:
- Auto-Type
- Browser Integration
- Legacy Browser Integration (KeePassHTTP)
- SSH Agent
- YubiKey

**Comment by droidmonkey** [comment]
Are you using KDBX4?

**Comment by mbway** [comment]
the database in question uses AES-KDF (KDBX 3.1). Also note that until recently the database has opened fine on mobile and I did not change any settings until the problem occurred.

**Comment by phoerious** [comment]
Check if your encryption is Chacha20, which may be the reason. I recommend you upgrade the KDF to Argon2 anyway.

**Comment by mbway** [comment]
it is AES 256-bit. I never changed the KDF because I created the database ages ago and never changed the settings.

**Comment by phoerious** [comment]
Then check your KDF round settings. The defaults are benchmarked to take a second on your PC. Your mobile may not manage that. But I still recommend Argon2 with sensible rounds and memory settings. There is not really a reason to stay compatible with KeePassX.

**Comment by mbway** [comment]
fair enough about KeePassX, I just tried opening with it to investigate if any of the applications would give more information about what exactly is 'wrong' with the database. The number of rounds is not the issue, to prove it I set the rounds to 200000 which is 100 times smaller than the the value for 1 second on my laptop (which isn't particularly powerful anyway). Opening in Keepass2Android gives the same generic error of "An error occurred: The file header is corrupted. Less data than expected could be read from the file."

database settings at the moment:
AES 256 bit
AES-KDF KDBX 3.1
200000 rounds

compression off

**Comment by mbway** [comment]
if I remove the keyfile I get a different error from Keepass2Android: `An error occurred: Specified argument was out of the range of valid values. Parameter name pb`

**Comment by phoerious** [comment]
Somethings wrong with your database. Your headers got messed up.

**Comment by mbway** [comment]
Current database settings:
no keyfile
compression enabled
AES 256-bit
Argon2 (KDBX 4)
8 rounds, 64MiB, 4 threads

Ok, good point about the transfer. I'm using an FTP server on the phone to make the transfer, didn't even think to question it because it has never failed before. I checked the hash and it didn't match! I changed the transfer mode to force binary mode and the database loaded correctly, so something about the database made it appear as ASCII?
